Lib/functools.py

```diff
@@ -273,43 +273,130 @@ def reduce(function, sequence, initial=_initial_missing):
 ### partial() argument application
 ################################################################################
 
+
+class __PlaceholderTypeBase:
+    """The type of the Placeholder singleton.
+
+    Used as a placeholder for partial arguments.
+    """
+    __instance = None
+    __slots__ = ()
+
+    def __new__(cls):
+        if cls.__instance is None:
+            cls.__instance = object.__new__(cls)
+        return cls.__instance
+
+    def __bool__(self):
+        raise TypeError("Placeholder should not be used in a boolean context")
+
+    def __repr__(self):
+        return 'Placeholder'
+
+    def __reduce__(self):
+        return 'Placeholder'
+
+def __placeholder_init_subclass__(cls, *args, **kwargs):
+    raise TypeError(f"type '{cls.__name__}' is not an acceptable base type")
+
+Placeholder = type(
+    'PlaceholderType',
+    (__PlaceholderTypeBase,),
+    {'__slots__': (), '__init_subclass__': __placeholder_init_subclass__}
+)()
+
+
+def _partial_prepare_new(cls, func, args, keywords):
+    if args:
+        if args[-1] is Placeholder:
+            raise TypeError("trailing Placeholders are not allowed")
+        phcount = args.count(Placeholder)
+    else:
+        phcount = 0
+    if isinstance(func, cls):
+        pto_args = func.args
+        pto_phcount = func.placeholder_count
+        # merge args with args of `func` which is `partial`
+        if pto_phcount and args:
+            tot_args = list(pto_args)
+            nargs = len(args)
+            pos = j = 0
+            end = nargs if nargs < pto_phcount else pto_phcount
+            while j < end:
+                pos = tot_args.index(Placeholder, pos)
+                tot_args[pos] = args[j]
+                pos += 1
+                j += 1
+            if pto_phcount < nargs:
+                tot_args.extend(args[pto_phcount:])
+            phcount += pto_phcount - end
+            args = tuple(tot_args)
+        else:
+            phcount += pto_phcount
+            args = func.args + args
+        keywords = {**func.keywords, **keywords}
+        func = func.func
+    return func, args, keywords, phcount
+
+
+def _partial_prepare_call(self, args, keywords):
+    pto_phcount = self.placeholder_count
+    pto_args = self.args
+    if pto_phcount:
+        n = len(args)
+        if n < pto_phcount:
+            raise TypeError(
+                "missing positional arguments "
+                "in 'partial' call; expected "
+                f"at least {pto_phcount}, got {n}")
+        pto_args = list(pto_args)
+        pos = j = 0
+        while j < pto_phcount:
+            pos = pto_args.index(Placeholder, pos)
+            pto_args[pos] = args[j]
+            pos += 1
+            j += 1
```

Comment:
This can be optimized if pre-calculate indices of the placeholders:
```py
for j, i in enum(self._placeholder_indices):
    pto_args[i] = args[j]
```
and in the constructor:
```py
self._placeholder_indices = tuple(i for i, arg in enumerate(args) if arg is Placeholder)
```

Reply:
If we are after efficiency of python implementation and adding `operator` import is acceptable I can swap it with the one I have in this PR's description. It is by far the most efficient one making use of `operator.itemgetter` to nicely align positionals.

I have a fully working implementation adapted to this class.

I tried to keep this simple so chose slightly more complex than the simplest one.

Reply:
`operator` is already imported in `collections`, so importing it here won't change anything.